### PR TITLE
Fixes retrieval of the latest snapshot index blob

### PR DESCRIPTION
**Pull request by abeyad**
This commit ensures that the index.latest blob is first examined to
determine the latest index-N blob id, before attempting to list all
index-N blobs and picking the blob with the highest N.

It also fixes the MockRepository#move so that tests are able to handle
non-atomic moves.  This is done by adding a special setting to the
MockRepository that requires the test to specify if it can handle
non-atomic moves.  If so, then the MockRepository#move operation will be
non-atomic to allow testing for against such repositories.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 1]
```diff
@@ -733,22 +733,21 @@ void writeIncompatibleSnapshots(RepositoryData repositoryData) throws IOExceptio
     long latestIndexBlobId() throws IOException {
         try {
             // first, try listing the blobs and determining which index blob is the latest
```
Comment (imotov) [inline core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: This comment is probably wrong now.

File: core/src/test/java/org/elasticsearch/snapshots/mockstore/MockRepository.java [hunk 2]
```diff
@@ -116,6 +116,7 @@ public MockRepository(RepositoryMetaData metadata, Environment environment,
         blockOnInitialization = metadata.settings().getAsBoolean("block_on_init", false);
         randomPrefix = metadata.settings().get("random", "default");
         waitAfterUnblock = metadata.settings().getAsLong("wait_after_unblock", 0L);
+        nonAtomicMove = metadata.settings().getAsBoolean("non_atomic_move", false);
```
Comment (imotov) [inline core/src/test/java/org/elasticsearch/snapshots/mockstore/MockRepository.java]: I think negatively named settings are somewhat difficult to reason about. For example, I think it's much easier to understand that `atomicMove = true` is than `nonAtomicMove = false`.

File: core/src/test/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java [hunk 3]
```diff
@@ -2738,6 +2738,7 @@ public void testSnapshotSucceedsAfterSnapshotFailure() throws Exception {
                 Settings.builder()
                     .put("location", repoPath)
                     .put("random_control_io_exception_rate", randomIntBetween(5, 20) / 100f)
+                    .put("non_atomic_move", true)
```
Comment (imotov) [inline core/src/test/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java]: I am not really sure why this test is singled out. Maybe add comment?

**Review by imotov — APPROVED** [review]
Left a couple of minor comments.

File: core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java (hunk 1 above)
Comment (abeyad) [inline core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: done

File: core/src/test/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java (hunk 3 above)
Comment (abeyad) [inline core/src/test/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java]: done

**Comment by abeyad** [comment]
5.x: ff0350d3c0714d48b864ebc197583ce730937e6d
5.2: 35f572901c639c7fea7531852d715787ea4591de

**Pull request merged**